Initial implementation

masylum opened the pull request:
#### :tophat: What? Why?

First prototype to support https://github.com/ipfs/specs/tree/master/repo
Implemented part of `fs-repo` adapter by using `fs-blob-store`, paving the way
for browser-compatible blob stores (@somebody32 working on it)

#### :ghost: GIF
![](https://media.giphy.com/media/fPY3NXWQwhmMg/giphy.gif)

^ streams

masylum commented:
@somebody32 I removed babel because it was slow and got tired of it, the code looks all-right anyway

masylum commented:
@diasdavid would be nice if you could take a look at the initial proposal to be sure I'm on the right way

masylum commented on src/adaptors/fs.js:
```diff
@@ -0,0 +1,67 @@
+    var LOCK_PATH = this.LOCK_PATH
+
+    function onFinish (buff) {
+      lockFile.unlock(LOCK_PATH, function () {
```
missing error handling

diasdavid commented on src/adaptors/fs.js:
```diff
@@ -0,0 +1,88 @@
+    return this._write(key, content, 'a', cb)
+  },
+
+  write: function (key, content, cb) {
```
It would be nice to also offer a stream interface

masylum replied:
good point, initially I was returning the writeable stream but it got lost when implementing the locks. Will create a richer API as I move forward.

diasdavid commented on README.md:
````diff
@@ -1,2 +1,103 @@
+})
+```
+
+## Adaptors
````
Maybe keep calling them blob-stores and point to abstract-blob-store?

masylum replied:
the issue I'm having is that those adaptors can't just be blob-stores. There is some extra plumbing that needs to be done around. For instance the `fs-repo` need locking through a `repo.lock` file which may be implemented completely different on a `IndexedDB` or `localstorage` implementation. Also, I may want to list all the blocks in a repository, that's not something doable within a blob-store and will have different implementations.

diasdavid commented:
Seems you are in the right direction :) I like that the interface revolves the several types of data stored in the Repo. It would be good to start having tests for the experiments you are doing, so other people feel confident to PR too.

You had some thoughts on how repo should be structured on disk, might wanna share with @whyrusleeping before 0.4.0 release, since 0.4.0 has a different repo structure already, it will be the perfect time for other significant migrations.

masylum commented:
Is there anywhere I can see what's planned in the 0.4.0 release? I've reached a point where I will have to start reverse-engineering the go implementation though since specs and reality don't seem to match (opened https://github.com/ipfs/specs/issues/41)

jbenet commented:
@masylum I'm sorry about that, I really need to fix the spec there to
reflect go. :c

I think we can move the nesting to be 3 levels in go-ipfs but some other
stuff is better than I need to materialize here.

If anybody wants to help me writing this stuff up it would be sweet.
On Wed, Nov 18, 2015 at 04:01 Pau Ramon Revilla <notifications@github.com>
wrote:

> Is there anywhere I can see what's planned in the 0.4.0 release? I've
> reached a point where I will have to start reverse-engineering the go
> implementation though since specs and reality don't seem to match (opened
> ipfs/specs#41 <https://github.com/ipfs/specs/issues/41>)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipfs/js-ipfs-repo/pull/1#issuecomment-157690464>.
>

whyrusleeping commented:
tv left some notes here that may help: https://github.com/ipfs/go-ipfs/blob/master/repo/fsrepo/fsrepo.go#L356

masylum commented:
I moved forward with the prefix as it is in go-ipfs. If the spec materializes in some other way is easy for me to change it back.

Added a `memory-repo` dummy implementation which will allow me to start testing easily. It will also help me define a `fs` agnostic API (I still have my doubts whether `repo.lock` should belong to repo or fs-repo).

I also will need some guidance with the rest since I'm not sure I understand the `datastore` folder.

jbenet commented:
@masylum sounds good to me. 

re the repo.lock, i responded in the spec. gist is that yes, that probably should be removed from `repo` spec.

diasdavid commented:
@masylum you have write perms on this repo, mind if you do a PR to a branch in this repo and then a PR to master, so we can work together without having to PR the PR the PR :)

diasdavid commented:
I went ahead and moved the PR to be a branch on the main repo, so it enables easier collaboration - https://github.com/ipfs/js-ipfs-repo/pull/2. Closing this one.
